### PR TITLE
Enabling Connectivity Analyses in FIRRTL

### PR DESCRIPTION
**Pull request by azidar**
Previously, it was difficult to do analyses of the FIRRTL AST that required connectivity information. A good example of this is dead-code elimination, which required many custom datastructures in order to work properly. Due to this start-up complexity, a novice FIRRTL-transform writer could not easily write transforms which needed this feature.

This PR aims to lower the threshold for transform-writers to write powerful transformations that use connectivity information. As an example, this PR also includes a clock-domain analyses transform, which given annotated signals, determines which clock signal (or clock signals) is driving it.

LFC:

- `CircuitGraph`/`ConnectionGraph`/`IRLookup` software architecture
- More use cases for netlist-type analyses

This PR adds both infrastructure and a new FIRRTL feature.

- `ConnectionGraph`: A connectivity-graph of FIRRTL local targets, but virtualizes nodes as an instance-viewed node as the graph is walked. This enables connectivity analyses with a smaller memory footprint. Also, previously explored modules will cache input-output connections and shortcut future trips through the module
- `IRLookup`: A class that enables correlating AST information (declarations, types, gender, kind) with targets, like the ones used as nodes in the `ConnectionGraph`
- `ClockSourceFinder`: A subclass of `ConnectionGraph` that lets you determine, given a signal, what clock source is it synchronized with. Intermediate results are cached to speed up queries based on past queries. It also can be done hierarchically, which means that results from leaf modules can also speed up queries from parent modules.
- `FindClockSources`: A transform that, given `GetClockSources` annotations, produces `ClockSources` annotations. It also consumes earlier `ClockSources` annotations to speed up new analyses, as well as enabling users to provide their own mapping (e.g. associating top-level inputs with the top-level clock port)

TODO:

- [x] ScalaDoc CircuitGraph functions
- [ ] Test CircuitGraph functions
- [x] Write RFC for expression tags (add new Tag type, could be super class of reference strings, which is optional for DoPrim, Mux, and ValidIf (and other WExpressions like WInvalid)
- [x] Move UnnamedToken etc. to new expression tags PR
- [x] Try to abstract more of `ClockSourceFinder` into `ConnectionGraph`, specifically the `recordClock` functions.
- [ ] ScalaDoc ConnectionGraph functions
- [ ] Add tagNode functions to ConnectionGraph

### CODE REVIEW

File: src/test/scala/firrtlTests/transforms/ClockFinderTransformSpec.scala [hunk 1]
```diff
@@ -0,0 +1,876 @@
...
+
+    execute(
+      input,
+      Seq(GetClockSources(allSignals)),
+      ClockSources(clockSources),
+      None
+    )
+  }
+
+  "Data read from sequential memories" should "return clock domain of read/readwrite's enable and address" in {
```
Comment (albert-magyar) [inline src/test/scala/firrtlTests/transforms/ClockFinderTransformSpec.scala]: Why not the clock attached to the port itself?

File: src/main/scala/firrtl/analyses/ConnectionGraph.scala [hunk 2]
```diff
@@ -0,0 +1,632 @@
...
+      if (node.path.nonEmpty) {
+        recSetTag(root, node.stripHierarchy(1), perModuleTags, tagMap)
+      }
+    }
+    recSetTag(node.module, node, perModuleTags, tagMap)
+  }
+
+  private def updatePerModuleTags(tag: ReferenceTarget, perModuleTags: mutable.HashMap[String, mutable.HashSet[ReferenceTarget]]): Unit = {
+    perModuleTags.getOrElseUpdate(tag.module, mutable.HashSet.empty[ReferenceTarget]) += tag
+    if(tag.path.nonEmpty) {
```
Comment (albert-magyar) [inline src/main/scala/firrtl/analyses/ConnectionGraph.scala]: `if (`

File: src/main/scala/firrtl/analyses/ConnectionGraph.scala [hunk 3]
```diff
@@ -0,0 +1,632 @@
...
+      case TokenTagger.literalRegex(value) => true
+      case _ => false
+    }
+  }
+
+  /** Within a module, given an [[Expression]] inside a module, return a corresponding [[Target]]
+    * @param m Target of module containing the expression
+    * @param tagger Used to uniquely identify unnamed targets, e.g. primops
+    * @param e
+    * @return
```
Comment (albert-magyar) [inline src/main/scala/firrtl/analyses/ConnectionGraph.scala]: fill in or delete return annotation

File: src/main/scala/firrtl/analyses/ConnectionGraph.scala [hunk 4]
```diff
@@ -0,0 +1,632 @@
...
+  override def findSCCs: Seq[Seq[ReferenceTarget]] = Utils.throwInternalError("Cannot call findSCCs on ConnectionGraph")
+}
+
+object ConnectionGraph {
+
+  /** Returns a [[DiGraph]] of [[Target]] and corresponding [[IRLookup]]
+    *
+    * Represents the directed connectivity of a FIRRTL circuit
+    * @param circuit
+    * @return
```
Comment (albert-magyar) [inline src/main/scala/firrtl/analyses/ConnectionGraph.scala]: fill in or delete return annotation

File: src/main/scala/firrtl/transforms/clockfinder/ClockFinder.scala [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+  /** Returns instance-viewed combinational-edges or reg-to-clock edges
+    * Ends early if visiting a node that was previously visited in another BFS
+    * @param node the specified node
+    * @param prevOpt
+    * @return a Set[T] of all vertices that source has edges to
+    */
+  override def getEdges(node: ReferenceTarget,
+                        prevOpt: Option[collection.Map[ReferenceTarget, ReferenceTarget]]
+                       ): collection.Set[ReferenceTarget] = {
+    val prev = prevOpt.get
```
Comment (albert-magyar) [inline src/main/scala/firrtl/transforms/clockfinder/ClockFinder.scala]: exposes `NoSuchElementException`

**Comment by ekiwi-sifive** [comment]
Can we close this now that https://github.com/freechipsproject/firrtl/pull/1603 is merged?

**Comment by sequencer** [comment]
close since #1603 got merged. But still notice that we are missing the clock domain analysis in the originally PR. Which I’ll pick or reimplement in my next PRs.

**Pull request closed without merging**